hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,717 @@
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+    ret = nrfx_qspi_lfm_xfer(buf, nullptr, size, true); // Send data to write, finalize long frame transfer
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+
+    return 0;
+}
+
+int erase_common(uintptr_t start_addr, size_t num_blocks, nrf_qspi_erase_len_t len) {
+    FlashLock lk;
+
+    int err_code = NRF_SUCCESS;
+
+    const size_t block_length = len == NRF_QSPI_ERASE_LEN_4KB ? 4096 : 64 * 1024;
+
+    /* Round down to the nearest block */
+    start_addr = ((start_addr / block_length) * block_length);
+
+    for (size_t i = 0; i < num_blocks; i++) {
+        err_code = nrfx_qspi_erase(len, start_addr);
+        CHECK_TRUE(err_code == NRFX_SUCCESS, nrf_system_error(err_code));
+
+        exflash_qspi_wait_completion();
+
+        start_addr += block_length;
+    }
+
+    // LOG_DEBUG(TRACE, "Erased %lu %lukB blocks starting from %" PRIxPTR,
+    //           num_blocks, block_length / 1024, start_addr);
+
+    return 0;
+}
+
+} // Anonymous
+
+int hal_exflash_init(void) {
+    // PATCH: Initialize CS pin, external memory discharge
+    nrf_gpio_cfg_output(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_pin_set(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_cfg_input(QSPI_FLASH_IO1_PIN, NRF_GPIO_PIN_PULLDOWN);
+
+    FlashLock lk;
+
+    // Use single data line SPI to read ID
+    nrfx_qspi_config_t default_config = NRF_QSPI_QUICK_CFG(NRF_QSPI_WRITEOC_PP, NRF_QSPI_READOC_FASTREAD, NRF_QSPI_FREQ_32MDIV2);
+    int ret = nrfx_qspi_init(&default_config, nullptr, nullptr);
+    if (ret) {
+        if (ret == NRFX_ERROR_TIMEOUT) {
+            // Could have timed out because the flash chip might be in an ongoing program/erase operation.
+            // Suspend it.
+            LOG_DEBUG(TRACE, "QSPI NRFX timeout error. Suspend pgm/ers op.");
+            ret = hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SUSPEND_PGMERS, nullptr, nullptr, 0);
+        }
+        return nrf_system_error(ret);
+    }
+
+    // Determine chip type / wake up from potential sleep
+    uint8_t chip_id[3] = {};
+    nrf_qspi_cinstr_conf_t read_id_command = {
+        .opcode    = HAL_QSPI_CMD_STD_READ_ID,
+        .length    = NRF_QSPI_CINSTR_LEN_4B,
+        .io2_level = true,
+        .io3_level = true,
+        .wipwait   = true,
+        .wren      = true
+    };
+    
+    // If read chip_id fails, it may be a GD25 that is asleep
+    if(!exflash_qspi_cinstr_xfer(&read_id_command, nullptr, chip_id))
+    {
+        nrf_qspi_cinstr_conf_t wake_command = {
+            .opcode    = HAL_QSPI_CMD_GD25_WAKE,
+            .length    = NRF_QSPI_CINSTR_LEN_1B,
+            .io2_level = true,
+            .io3_level = true,
+            .wipwait   = true,
+            .wren      = true
+        };
+        CHECK(exflash_qspi_cinstr_xfer(&wake_command, nullptr, NULL));
```

Comment:
Nit/Style: nullptr

Reply:
A 20us delay is needed before reading ID to ensure the flash is woken up.

![Screen Shot 2022-06-13 at 10 42 19](https://user-images.githubusercontent.com/7424522/173270235-3354c487-a79c-40f2-9dd9-2d1c8018afc9.png)